test/jdk/java/net/Socket/CtorFailTest.java

```diff
@@ -0,0 +1,229 @@
+            });
+
+            // Run verifications
+            Exception expectedException = socketImpl.bindException != null
+                    ? socketImpl.bindException
+                    : socketImpl.connectException;
+            assertSame(expectedException, caughtException);
+            assertEquals(1, socketImpl.closeInvocationCounter.get());
+
+        });
+    }
+
+    static List<MockSocketImpl> socketImpls() {
+        String exceptionMessage = "intentional test failure";
+        IOException checkedException = new IOException(exceptionMessage);
+        IllegalArgumentException uncheckedException = new IllegalArgumentException(exceptionMessage);
+        return List.of(
+                new MockSocketImpl(checkedException, null),
+                new MockSocketImpl(null, checkedException),
+                new MockSocketImpl(uncheckedException, null),
+                new MockSocketImpl(null, uncheckedException));
+    }
+
+    private static final class MockSocketImpl extends SocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindException;
+
+        private final Exception connectException;
+
+        private MockSocketImpl(Exception bindException, Exception connectException) {
+            this.bindException = bindException;
+            this.connectException = connectException;
+        }
+
+        @Override
+        protected void create(boolean stream) {
+            // Do nothing
+        }
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindException);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectException);
+        }
+
+        private void throwIfPresent(Exception exception) throws IOException {
+            if (exception != null) {
+                switch (exception) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown exception type: " + exception.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        public String toString() {
+            Map<String, Object> valueByKey = new LinkedHashMap<>();
+            if (bindException != null) {
+                valueByKey.put("bindException", bindException.getClass().getSimpleName());
+            }
+            if (connectException != null) {
+                valueByKey.put("connectException", connectException.getClass().getSimpleName());
+            }
+            valueByKey.put("closeInvocationCounter", closeInvocationCounter.get());
+            return MockSocketImpl.class.getSimpleName() + valueByKey;
```

Comment:
Have you considered using record instead of Map? You can create a local record class within a method...

```
        record MockSocket(Exception bindException, Exception connectException) { }
        return new MockSocket(bindException, connectException).toString();
```

Reply:
Switched to using records.